### DISCUSSION TITLE
UseItems composable loading ref

### DISCUSSION THREAD

**Discussion by MichaelFBA** [directus/directus#16290] [Archive]
When using the `useItems` composable in the directus extensions SDK the `loading` reference is set to false. This causes issues since the `items` is undefined when it is instantiated and we cannot rely on `loading` to conditionally render a component without adding some nasty watcher on `items`

### To Reproduce

In the example below `<SomeComponent>` will render and items will be initially empty until it eventually returns data, causing issues further down component tree
```
<template>
   <div v-if="loading">Loading</div>
   <SomeComponent v-else :items="items"/> 
</template>

const { items, loading } = useItems(ref('collection_name'), {
    fields: ref(['*']),
    limit: ref(-1),
    sort: ref(null),
    search: ref(''),
    filter: ref(null),
    page: ref(1)
});
```

**Comment by rijkvanzanten** [comment]
This is working as intended. The items aren't being loaded immediately, once the actual loading starts, the `loading` ref is set to true 👍🏻

**Reply by MichaelFBA** [reply]
Hi @rijkvanzanten 
I will take a closer look at debugging the useItems component later today, I did notice loading is wrapped in a setTimeout, not sure why that is?
The issue I am seeing is if I have a component prop that uses items.id for example it will throw and error because there is a race condition before items and loading is set

I can create an example if need be

**Reply by MichaelFBA** [reply]
https://github.com/directus/directus/blob/5fe28db539fcac5dfd6a529dc78338fe32d1d475/packages/shared/src/composables/use-items.ts#L114

Removing this set timeout would fix the issue I think.

**Reply by rijkvanzanten** [reply]
Ahh that part. Yeah that's there because we don't want to show the spinner UX for item loads that take less than 150ms. Down to discuss optimizations tho! 🙂

**Reply by MichaelFBA** [reply]
Im not sure how the spinner ux is related TBH. But in its current state the loading ref is pretty unusable. I have basically recreated useItems without the timeout to get around the functionality.
Maybe if you provide an example of how the spinner ux works I could make a suggestion. 

This is my use case as above
```
<div v-if="loading">Loading</div>
<SomeComponentThatRequiresItems v-else :items="items"/
```

**Reply by azrikahar** [reply]
I believe the UX here is in regards to mitigate abrupt loading/spinner flashes when we're filtering/searching on every keystroke. [This tweet](https://twitter.com/MichaelThiessen/status/1548748122700107777) also describes this particular approach. (fully aware that a tweet does not and should not constitute as the single source of truth, but it's just coincidentally the one that I recall coming across off the top of my mind) 

Seems like the layouts themselves are currently handling it by checking items count and display no-results/no-items slot accordingly. Here are some examples from Cards and Table layout respectively:

- Cards

    https://github.com/directus/directus/blob/c70211c48dee133dbf7f19785dcbf5ea9bba158c/app/src/layouts/cards/cards.vue#L3
    
    https://github.com/directus/directus/blob/c70211c48dee133dbf7f19785dcbf5ea9bba158c/app/src/layouts/cards/cards.vue#L72-L73

- Table

    https://github.com/directus/directus/blob/c70211c48dee133dbf7f19785dcbf5ea9bba158c/app/src/layouts/tabular/tabular.vue#L3-L4
    
    https://github.com/directus/directus/blob/c70211c48dee133dbf7f19785dcbf5ea9bba158c/app/src/layouts/tabular/tabular.vue#L169-L170
    
Note that this by no means meant to suggest that your setup should follow these, but more so to share the relevant findings here :thumbsup:

With that said, I can totally understand your expected loading pattern/behavior! Rather than removing it, perhaps there can be a prop that defaults to true which uses setTimeout, so that users can explicitly disable it while keeping it backward compatible. Alternatively maybe it can be a number prop `delay` similar to what [Vue's `defineAsyncComponent`](https://vuejs.org/guide/components/async.html#loading-and-error-states) has, which coincidentally has an excerpt mentioning the general idea behind such delay that is also applicable here:

>  this is because on fast networks, an instant loading state may get replaced too fast and end up looking like a flicker.

**Reply by MichaelFBA** [reply]
Makes sense thanks for the explanation @azrikahar 

I think the optional prop `delay?=150` sounds like a good idea 👍 

In other news I also noticed the other optional prop fetchOnInit defaults to true which in my case makes 2 subsequent requests to the same endpoint since the watch also has `immediate:true`. Not sure if this is intended?

**Reply by azrikahar** [reply]
> In other news I also noticed the other optional prop fetchOnInit defaults to true which in my case makes 2 subsequent requests to the same endpoint since the watch also has `immediate:true`. Not sure if this is intended?

Interesting! From what I was able to gather, it seems like it used to be `immediate: fetchOnInit` <https://github.com/directus/directus/pull/2893/files#diff-0d53067b4a1ccc2bb89ebeb8bfc7bbda1c7f94922e75668b223feaa0c64741b5R62>, however use-items was refactored in #8570 and that ended up being changed 🤔 My current assumption would be it could be intended due to particular layouts, but I would also agree that it seems odd and may require further investigation to confirm whether it's still valid, or should it be "only enabled for certain layouts" if removing it will affect any existing layouts.

**Reply by MichaelFBA** [reply]
The immediate: fetchOnInit makes sense so it doesn't  trigger an initial request. 
I'm not sure if you have tests in place for the various layouts? But perhaps it would be a nice edition

**Reply by azrikahar** [reply]
> I'm not sure if you have tests in place for the various layouts? But perhaps it would be a nice edition

There's no layout-specific tests yet (we are actively improving tests!), but I do agree that it would be great, or rather pivotal, for better stability when we're looking to go forward with changes such as this.
